### PR TITLE
[FLINK-35313][mysql-cdc] Add upsert changelog mode to avoid UPDATE_BEFORE records …

### PR DESCRIPTION
**Pull request by yeezychao**
This PR implements [#1898](https://github.com/ververica/flink-cdc-connectors/issues/1898) and exposing support in configuration just for mysql module.

### CODE REVIEW

**Comment by yuxiqian** [comment]
Hi @yeezychao, thanks for your great contribution! Before this PR could be merged, could you please rebase it with latest `master` branch?

**Comment by yeezychao** [comment]
Hi @yuxiqian, thanks for your reply, PTAL.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java
Comment (yuxiqian) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/table/MySqlTableSourceFactory.java:1]: Is this file content modified or it's just the metadata change?

**Review by yuxiqian — APPROVED** [review]
Looks good to me, could @PatrickRen please trigger the CI please?

**Comment by yuxiqian** [comment]
Hi @yeezychao, seems MySQL CI job is failing, could you please take a look? https://github.com/apache/flink-cdc/actions/runs/8844257194/job/24756979040?pr=1907

(The markdown lint check is irrelevant to this PR and should be fixed with a `git rebase master`.)

**Comment by yeezychao** [comment]
> Hi @yeezychao, seems MySQL CI job is failing, could you please take a look? https://github.com/apache/flink-cdc/actions/runs/8844257194/job/24756979040?pr=1907
> 
> (The markdown lint check issue is irrelevant to this PR and should be fixed with a `git rebase master`.)

@yuxiqian Fixed

**Comment by yeezychao** [comment]
> Hi @yeezychao, seems MySQL CI job is failing, could you please take a look? https://github.com/apache/flink-cdc/actions/runs/8844257194/job/24756979040?pr=1907
> 
> (The markdown lint check issue is irrelevant to this PR and should be fixed with a `git rebase master`.)

@yuxiqian Fixed

**Comment by yuxiqian** [comment]
Hi @yeezychao, you may run `mvn spotless:apply` to fix style violations, and use `mvn verify` to ensure following the style rules.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/table/MySqlTableSourceFactory.java [hunk 1]
```diff
@@ -47,6 +48,8 @@
 import java.util.Set;
 import java.util.regex.Pattern;
 
+import static org.apache.flink.cdc.common.utils.Preconditions.checkArgument;
+import static org.apache.flink.cdc.connectors.mysql.source.config.MySqlSourceOptions.CHANGELOG_MODE;
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/table/MySqlTableSourceFactory.java:52]: Maybe use qualified `MySqlSourceOptions.CHANGELOG_MODE` for consistency?

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/table/MySqlTableSourceFactory.java [hunk 2]
```diff
@@ -189,6 +198,7 @@ public Set<ConfigOption<?>> optionalOptions() {
         options.add(MySqlSourceOptions.HEARTBEAT_INTERVAL);
         options.add(MySqlSourceOptions.SCAN_INCREMENTAL_SNAPSHOT_CHUNK_KEY_COLUMN);
         options.add(MySqlSourceOptions.SCAN_INCREMENTAL_SNAPSHOT_BACKFILL_SKIP);
+        options.add(CHANGELOG_MODE);
```
Comment (yuxiqian) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/table/MySqlTableSourceFactory.java:201]:
```suggestion
        options.add(MySqlSourceOptions.CHANGELOG_MODE);
```

**Comment by yeezychao** [comment]
@yuxiqian Turning on upsert mode still fails to filter -u data. I am very confused as to why the same PR application failed to test in cdc 3.2 (flink 1.18) version, but it still works in version 2.2 (flink 1.15). Unfortunately, I have not found the reason yet.
image

**Comment by yuxiqian** [comment]
> @yuxiqian Turning on upsert mode still fails to filter -u data. I am very confused as to why the same PR application failed to test in cdc 3.2 (flink 1.18) version, but it still works in version 2.2 (flink 1.15). Unfortunately, I have not found the reason yet. image

Maybe checking the printed log and confirm if MySQL source actually sends any `-U` events to downstream? IIRC Flink will automatically append a `ChangelogNormalize` node to backfill missing update before events if source doesn't provide it.

**Comment by yeezychao** [comment]
> > @yuxiqian Turning on upsert mode still fails to filter -u data. I am very confused as to why the same PR application failed to test in cdc 3.2 (flink 1.18) version, but it still works in version 2.2 (flink 1.15). Unfortunately, I have not found the reason yet. image
> 
> @yeezychao Maybe check the output log and confirm if MySQL source actually sends any `-U` events to downstream? IIRC Flink will automatically append a `ChangelogNormalize` node to backfill missing update before events if source doesn't provide it.

You are right！The `ChangelogNormalize` node is indeed added under the Flink 1.18 version，but the Flink 1.15 version haven't.
image

**Pull request closed without merging**